### ISSUE TITLE
Rewrite the media processing pipeline

### ISSUE THREAD

**Issue by ungive** [ungive/discord-music-presence#84]
Right now everything's a bit all over the place with how media is handled internally, from detection, to sanitizing it, fixing badly formatted metadata from some media players, making API requests, to getting a URL for a locally available cover image and more. This needs to be reworked such that that media processing pipeline is well-structured and can be extended with more than just one media source (system media) and more than just one entity that consumes the media (the Discord status).

Reworking this lays the foundation for:
- Getting media from additional sources, like web browsers, next to system media alone
- Consuming media for more than just one purpose, e.g. scrobbling, next to just a Discord status
- Improving maintainability and extensibility for the future and reducing code complexity
- Less code redundancy with system media detection
- Improved performance

This is required for #53, #5 and #62.

Mandatory todos:
- [x] Implement data structures for a media pipeline including different types of hooks that can manipulate and work with media in the pipeline and that can be registered at different points.
- [x] Implement a universal algorithm to detect changes in media properties between two media instances which compares media based on a selectable subset of properties (e.g. only title and artist and/or other properties)
- [x] Write tests to ensure existing pipeline data structures work as intended
- [x] Modify the Media data structure to have any number of unique source identifiers (platform/player ID/disambiguation tuples, collectively called "source ID") and optional source metadata (player display name) per identifier and implement comparison and hashing functions so that the source ID of each Media instance can be used as map keys
- [x] Modify the underlying system media detection code so it uses the new media source data structures
- [x] Modify the Linux media detection code to report two source IDs for media from browsers, the MPRIS player ID and the domain name that is retrieved from the MPRIS url field, in order to test that multiple source IDs work correctly and make sure that websites are identified properly
- [x] Now implement change detection between consecutive pipeline runs, where media from the first pipeline run is compared with media from the second pipeline run in such a way, that media with the same source ID is compared to each other for equality. This is only possible after implementing the above source ID type
- [x] Allow each consumer (Discord status, scrobbling, etc.) to determine the attributes that need to be compared for changes (using the comparison algorithm from before) for each media instance that passes through the pipeline (i.e. they select which attributes are important to them at any given point in time and per media instance). That way the Discord status consumer can e.g. control that comparing playing media involves the current playback position of the song, but paused media does not (since showing paused media in the status never shows the current playback position). This abstracts away change detection from the consumers and makes the pipeline a lot more efficient since consumers are only executed when the data they care about actually changed and additional expensive operations, like calls to music APIs are only made when really needed.
- [x] Fix the pipeline to have a copy of the processed media for each consumer and for selectable amender hooks to operate on these copies. Fix amenders selected by one consumer modifying media for another consumer
- [x] Run each consumer on a separate thread and join those threads during a pipeline run
- [x] Allow retrieving all registered hooks from the pipeline which will be very useful when embedding toggle checkboxes for them in the UI. That way transformer hooks like the fix for Apple Music do not need to be hardcoded the same way as before
- [x] Allow to disable each registered pipeline hook individually
- [x] Allow consumer-specific hooks that are executed before media/consumer/amender selection for consumers that are selected using the API of the pipeline. Unlike amenders these hooks are not selected by consumers but by the user of the pipeline which allows certain hooks (e.g. only allow podcasts from Spotify) to be shown for certain consumers only (e.g. only the Discord status). "only allow podcasts from Spotify" doesn't make sense for scrobbling e.g.
- [x] Add the ability to manually fire pipeline runs with current media at any given point in time
- [x] Allow any hook to define a callback when it is enabled or disabled
- [x] Treat all hooks as disabled when all consumers are disabled and no consumers are executed on a pipeline run
- [x] Make all calls into a hook implicitly thread-safe (by handling synchronization in the pipeline, not in the hook's code)
- [x] Allow cancellation of pipeline runs, which is desirable when a new media update is coming in and a previous pipeline run is still active. Especially in regard to requests to music APIs, where long-running requests might completely block further media processing. This requires the API client to offer a way to cancel an ongoing request.
- [x] Allow hooks to have a thread-safe options object that can be used to configure the hook. The hook should be able to get real-time updates whenever the option object is updated. Hooks should receive a callback when the options are updated. Hooks should not have to implement thread synchronization at all, instead it is the responsibility of the pipeline to synchronize calls into a hook. This is important, since hook callbacks and pipeline runs can happen in parallel.
- [x] Do not configure a hook nor call the configured callback when the configuration has not changed at all.
- [x] When a modifier or amender hook is configured, fire a pipeline run. Only fire the pipeline run after all concurrent hook configure calls have completed (it makes little sense to fire a pipeline run after every singular configure operation). Now, if the configuration of a modifier or amender changes the handled pipeline media metadata meaningfully, consumers should be executed again, given that they selected the attributes that changed meaningfully.
- [x] Fix consumers executing repeatedly when the media has no changes and a transformer or amender has modified the media in the same way consistently (compare media right before executing the consumer)
- [x] Migrate all "amenders" (components that add media metadata using external sources, e.g. the TIDAL API)
- [x] Fix long-running amenders being fully executed before any consumer is executed. Instead execute the consumer as soon as all of its selected amenders have finished running. Currently all consumers are only executed when all selected amenders have finished. In practice this would e.g. block the scrobbling consumer until e.g. an API request for the Discord consumer has completed, which would introduce unnecessary delays.
- [x] Fix amenders executing even when consumers would dismiss the media due to no changes in the observed attributes. This executes amenders more often than they need to be.
- [x] Fix consumer execution history piling up and using up more and more memory over time (albeit slowly) in rare cases where the OS reports many different media players (e.g. Linux with its many "chromium.instance2451" players). This is trivial to implement.
- [x] Fix hooks being configurable *during* a pipeline run, which is honestly undesirable. Configuration should not change while the individual hook methods ("media", "attributes", "operator()") are executed. That makes little sense and can introduce subtle bugs where the hook options are configured to have different values in-between one of the steps.
- [x] Fix consumer not executing during pipeline run 3, when it executed with media A during pipeline run 1, then pipeline run 2 filtered out media A and didn't execute the consumer and pipeline run 3 did not filter out media A anymore, while media A has not experienced any metadata changes between any runs. The consumer should execute during the third run, even though no metadata changed, since media A was not part of the second pipeline run.
- [x] Migrate the Discord status to be a pipeline consumer and ensure it is working well, perhaps write automated tests
- [x] Migrate all filters/transformers to the pipeline (e.g. Spotify ads filters, Apple Music fixes, etc.)
- [x] ~~Implement comprehensive logging of pipeline runs and execution of individual hooks, such that it's possible to very clearly make out how media has been modified exactly before it is used by a consumer.~~ Implemented basic logging, as comprehensive logging is really just a nice-to-have
- [x] ~~Refactor system media detection to use the universal comparison algorithm to replace the previous copy-pasted and redundant comparison code (which is identical on all platforms) - not mandatory for things to work, but reduces redundancy and can improve performance slightly by detecting changes more robustly and with less false-positives~~ This is moved into the pipeline. Media sources report media as frequently as they want to and media is compared against the history before being handled by the pipeline.
- [x] Complete!

**Comment by ungive** [comment]
The pipeline implementation is feature-complete. A beta build is available on the Discord server for beta testers!

**Comment by ungive** [comment]
I'm proud to say that the work on this is finally complete! The changes have proven to improve the stability, reliability and efficiency of the app by quite a lot and I'm very happy with how modular and maintainable the media processing components have become now, especially the Discord status. That has become a mess before. Working on future features will be a lot more straightforward.

Thank you to everyone on the Discord server who volunteered to test and tried to break the preview of version 2.3.0. While there weren't any significant bugs with the pipeline itself, it's surfaced quite a few other bugs in other places! The latest release has been out for a while and the stability of the app appears to be rock-solid so far! This wouldn't have been possible without you. :)

Closing. 🎉
